src/ring_theory/adjoin/basic.lean

```diff
@@ -55,6 +55,13 @@ le_antisymm (adjoin_le h₁) h₂
 theorem adjoin_eq (S : subalgebra R A) : adjoin R ↑S = S :=
 adjoin_eq_of_le _ (set.subset.refl _) subset_adjoin
 
+theorem adjoin_induction {p : A → Prop} {x : A} (h : x ∈ adjoin R s) (Hs : ∀ x ∈ s, p x)
+  (Halg : ∀ r, p (algebra_map R A r))
+  (Hadd : ∀ x y, p x → p y → p (x + y))
+  (Hmul : ∀ x y, p x → p y → p (x * y)) : p x :=
+(@adjoin_le _ _ _ _ _ _ ⟨p, ring_hom.map_one (algebra_map R A) ▸ Halg 1, Hmul,
+  ring_hom.map_zero (algebra_map R A) ▸ Halg 0, Hadd, Halg⟩) Hs _ h
```

Comment:
I think you can probably avoid the `map_zero` and `map_one` if you use `{}` instead of `⟨⟩`

Reply:
It's an extra line, but there's less going on:
```suggestion
let S : subalgebra R A :=
{ carrier := p, mul_mem' := Hmul, add_mem' := Hadd, algebra_map_mem' := Halg } in
adjoin_le (show s ≤ S, from Hs) h
```

Reply:
You mean just `(@adjoin_le _ _ _ _ _ _ {p, Halg 1, Hmul, Halg 0, Hadd, Halg}) Hs _ h`? This doesn't work, but indeed
`(@adjoin_le _ _ _ _ _ _ {p, ring_hom.map_one (algebra_map R A) ▸ Halg 1, Hmul,
  ring_hom.map_zero (algebra_map R A) ▸ Halg 0, Hadd, Halg}) Hs _ h` doesn't work, so I am probably missing something.

Reply:
I think we crossposted, see my comment above.